packages/taiko-client/driver/chain_syncer/blob/soft_block.go

```diff
@@ -0,0 +1,291 @@
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/encoding"
+	softblocks "github.com/taikoxyz/taiko-mono/packages/taiko-client/driver/soft_blocks"
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/internal/utils"
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/pkg/rpc"
+)
+
+// InsertSoftBlockFromTransactionsBatch inserts a soft block into the L2 execution engine's blockchain
+// from the given transactions batch.
+func (s *Syncer) InsertSoftBlockFromTransactionsBatch(
+	ctx context.Context,
+	blockID uint64,
+	batchID uint64,
+	txListBytes []byte,
+	batchMarker softblocks.TransactionBatchMarker,
+	blockParams *softblocks.SoftBlockParams,
+) (*types.Header, error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	parent, err := s.rpc.L2.HeaderByNumber(ctx, new(big.Int).Sub(new(big.Int).SetUint64(blockID), common.Big1))
+	if err != nil {
+		return nil, err
+	}
+
+	if parent.Number.Uint64()+1 != blockID {
+		return nil, fmt.Errorf("parent block number (%d) is not equal to blockID - 1 (%d)", parent.Number.Uint64(), blockID)
+	}
+
+	// Calculate the other block parameters
+	difficultyHashPaylaod, err := encoding.EncodeDifficultyCalcutionParams(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode `block.difficulty` calculation parameters: %w", err)
+	}
+	protocolConfigs, err := rpc.GetProtocolConfigs(s.rpc.TaikoL1, &bind.CallOpts{Context: ctx})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch protocol configs: %w", err)
+	}
+
+	var (
+		txList     []*types.Transaction
+		fc         = &engine.ForkchoiceStateV1{HeadBlockHash: parent.Hash()}
+		difficulty = crypto.Keccak256Hash(difficultyHashPaylaod)
+		extraData  = encoding.EncodeBaseFeeConfig(&protocolConfigs.BaseFeeConfig)
+	)
+
+	if err := rlp.DecodeBytes(txListBytes, &txList); err != nil {
+		return nil, fmt.Errorf("failed to RLP decode txList bytes: %w", err)
+	}
+
+	baseFee, err := s.rpc.CalculateBaseFee(
+		ctx,
+		parent,
+		new(big.Int).SetUint64(blockParams.AnchorBlockID),
+		true,
+		&protocolConfigs.BaseFeeConfig,
+		blockParams.Timestamp,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate base fee: %w", err)
+	}
+
+	// Insert the anchor transaction at the head of the transactions list.
+	if batchID == 0 {
+		// Assemble a TaikoL2.anchorV2 transaction.
+		anchorTx, err := s.anchorConstructor.AssembleAnchorV2Tx(
+			ctx,
+			new(big.Int).SetUint64(blockParams.AnchorBlockID),
+			blockParams.AnchorStateRoot,
+			parent.GasUsed,
+			&protocolConfigs.BaseFeeConfig,
+			new(big.Int).SetUint64(blockID),
+			baseFee,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create TaikoL2.anchorV2 transaction: %w", err)
+		}
+
+		txList = append([]*types.Transaction{anchorTx}, txList...)
```

Comment:
```suggestion
		txList = append(txList, anchorTx)
```

Reply:
Anchor tx must be the first one of each block